get_devices/web_app.js

```diff
@@ -0,0 +1,135 @@
+ *  limitations under the License.
+ */
+
+// Espruino devices publish a UART service by default.
+const nordicUARTService = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
+// The test service defined in device_code.js
+const testService = '0b30acec-193e-11eb-adc1-0242ac120002';
+// The test characteristic defined in device_code.js
+const testCharacteristic = '0b30afd0-193e-11eb-adc1-0242ac120002';
+
+let gattServer = undefined;
+
+/**
+ * Load the device code to the Espruino IDE.
+ */
+function loadEspruinoDeviceCode() {
+  fetch('device_code.js').then(response => response.text()).then(data => {
+    let url = 'http://www.espruino.com/webide?code=' + encodeURIComponent(data);
+    window.open(url, '_window');
+  });
+}
+
+function onGattDisconnected(evt) {
+  const device = evt.target;
+  logInfo(`Disconnected from GATT on device ${device.name}.`);
+  assertFalse(gattServer.connected, 'Server connected');
+}
+
+async function startPairing() {
+  clearStatus();
+  logInfo('Starting test');
+
+  logInfo(`Requesting Bluetooth device with service ${testService}`);
+
+  $('btn_load_code').disabled = true;
+  $('btn_start_pairing').disabled = true;
+  $('btn_refresh_page').disabled = true;
+  $('btn_check_permission').disabled = true;
+
+  try {
+    const options = {
+      filters: [
+        { services: [nordicUARTService] }
+      ],
+      optionalServices: [testService]
+    };
+    logInfo(`Requesting Bluetooth device with service ${testService}`);
+    const device = await navigator.bluetooth.requestDevice(options);
+    logInfo(`Connected to Bluetooh device with id: ${device.id} and name: ${device.name}`);
+
+    device.addEventListener('gattserverdisconnected', onGattDisconnected);
+    logInfo(`Connecting to GATT server for device \"${device.name}\"...`);
+    gattServer = await device.gatt.connect();
+    assertEquals(gattServer.device, device, 'Server device mismatch');
+    assertTrue(gattServer.connected, 'server.connected should be true');
+
+    logInfo(`Connected to GATT, requesting service: ${testService}...`);
+    const service = await gattServer.getPrimaryService(testService);
+    assertEquals(service.device, device, 'service device mismatch');
+
+    logInfo(`Connected to service uuid: ${service.uuid}, primary:${service.isPrimary}`);
+  } catch (error) {
+    logError(`Unexpected failure: ${error}`);
+  }
+
+  $('btn_load_code').disabled = false;
+  $('btn_start_pairing').disabled = false;
+  $('btn_refresh_page').disabled = false;
+  $('btn_check_permission').disabled = false;
+}
+
+async function refreshPage() {
+  logInfo('Refreshing the page')
+  location.reload();
+}
+
+async function checkPermission() {
+  clearStatus();
+  logInfo(`Calling getDevices API`);
+  const name = 'getDevices_test';
+  const pairedDevices = await navigator.bluetooth.getDevices();
+
+  var passed = false;
+  for (let i = 0; (i < pairedDevices.length) && !passed; i++) {
+    passed = pairedDevices[i].name == name;
```

Comment:
Each `BluetoothDevice` has an `id` attribute which you can check for equality against the attributes of the device you originally got from `requestDevice()`.

Reply:
The only issue is that I won't have access to that after refresh, unless I permanently store it locally.